gridcapacity/capacity_analysis.py

```diff
@@ -210,24 +209,46 @@ def handle_empty_contingency_scenario(
 
     def buses_headroom(self) -> Headroom:
         """Return actual load and max additional PQ power in MVA for each bus"""
-        buses: Buses = Buses()
         console.print("Analysing headroom", style="blue")
+
+        generate, total = self.create_buses_headroom_generator()
+        headroom: List[BusHeadroom] = []
+
         with tqdm(
-            total=len(buses)
+            total=total
             if self._selected_buses_ids is None
             else len(self._selected_buses_ids),
             postfix=[{}],
         ) as progress:
-            PowerFlows.reset_count()
-            ViolationsStats.reset()
-            return tuple(
-                self.bus_headroom(bus, progress)
-                for bus in buses
-                if self._selected_buses_ids is None
-                or bus.number in self._selected_buses_ids
-            )
+            for bus_headroom, _ in generate():
+                progress.postfix[0]["bus_number"] = bus_headroom.bus.number
+                progress.postfix[0]["power_flows"] = PowerFlows.count
+                progress.update()
+                headroom.append(bus_headroom)
+
+        return tuple(headroom)
+
+    def create_buses_headroom_generator(self):
```

Comment:
```suggestion
    def create_buses_headroom_generator(self) -> Callable[Generator[tuple[BusHeadroom, int]]]:
```
I'm not sure about this one, please check.

Reply:
Suggested type is incorrect. 
The right one would be `tuple[() -> Generator[tuple[BusHeadroom, Any], Any, None], int]`
However it is correctly inferred from return value. I think there is no need to specify it explicitly.